### ISSUE TITLE
Automatic Exception Handler

### ISSUE THREAD

**Issue by dhiegobastos** [roadrunnerlenny/etlbox#41]
Hello,
I was reading about this mention below on doc and I was not able to understand If the framework capture exceptions during the etl and continue the flow for the next rows or if I need to catch the exceptions on my own.

"Know your errors: When exceptions are raised you get the exact line of code where your ETL stopped, including a hands-on description of the error."

Thanks.

**Comment by roadrunnerlenny** [comment]
Hi, 
Thank you for your interest in ETLBox.
Actually there is no special error handling implemented in ETLBox. The line in the doc was supposed to state that when you debug your code the exception will be thrown right away - so e.g. Visual Studio will show you the line of code where the exception was thrown. But there is no special exception handling implemented on top of that.

I can see that this statement can be misleading and will remove it from the documentation.

By the way, having some exception handling in place that would store erroneous records in an separate object or flow could be helpful.
What would be your expected or desired behavior from a library like ETLBox?

**Comment by bencassie** [comment]
Hi there,

Chipping in here as this is something I need to think about also. 

I think the expected behaviour of ETLBox would be to facilitate the catching of these exceptional rows and pass them to another data flow handler.

I would probably be looking to be able to set up LinkTo with a destination for any rows that throw exceptions. I will think on it more though.

**Comment by roadrunnerlenny** [comment]
Hi @bencassie ,

I like the idea of having some kind of "error" output that could be linked to another data flow especially handling erroneous rows. Probably a RowTransformation could have the `LinkErrorTo`method on top of that. It probably would contain only the wrong data - I think the exception message would be only logged. 

I can imaging this could be very helpful for a RowTransformation. It would basic have a try/catch block around the user action, and every time an exception would be catched it would send the record to the error output.

But I am not quite if this is applicable for all components. E.g. it wouldn't work for a Multicast, perhaps it could work for a MergeJoin, probably not for a Sort.... Any ideas here?

**Comment by bencassie** [comment]
Hi,

I've recently had the misfortune of working on some SSIS packages and I guess I was thinking along the lines of how MS handle row level errors (as i'm sure you're aware) - https://docs.microsoft.com/en-us/sql/integration-services/data-flow/error-handling-in-data?view=sql-server-ver15. The first paragraph does a good job of summarising our situation. 

I think you're correct, it really is only Sources/Transformations and Destinations that would need this linkage.

As in the SSIS approach, what we'd ideally want is to encapsulate the offending row in an object along with the error message / exception object. That way the author can deal with the error however they wish in the error destination.

**Comment by dhiegobastos** [comment]
@roadrunnerlenny @bencassie 
I appreciate your ideas for creating a new dataflow for records with errors. I believe it is also necessary to log the information and exceptions that have occurred making it possible to view the details on the console or using the etlboxlogviewer tool (https://github.com/roadrunnerlenny/etlboxlogviewer).

**Comment by roadrunnerlenny** [comment]
@dhiegobastos  I agree, an additionally logging makes sense. Thanks for your input!

To summarize: 
- Exceptions can occur when reading data (all sources), when transforming data (`RowTransformation` only) or when writing data (all destinations). 
- This exception should not stop the flow, but only affect the record that is processed
- error data could be send to a new data flow connected via the `LinkErrorsTo` - method 
- the error data should come with the exception message
- the error should also be logged

One problem that I can imagine is that most of the destinations try to write the data in batches - so if an error occurs during a batch insert, the whole batch will be "faulted" and send to error flow. 

Another thing is that you actually would need to connect every component that could create an error to the error data flow. The bigger the flow gets, the more components need to be connected.

What do you think?

**Comment by dhiegobastos** [comment]
@roadrunnerlenny For me it's perfect.

**Comment by bencassie** [comment]
Hi,

Apologies I have only just gotten back to this. I agree with the points you've made. It is unfortunate that on a batch insert the whole batch will fail, but I guess that's the nature of the bcp.

You are right that you could end up connecting a lot of components, having thought on it more, I still think it makes sense to keep the error handling logic away from the business logic. At least now that you can fluently connect the flows it makes it easier to add the error sinks.

Thanks

**Comment by roadrunnerlenny** [comment]
Hi @bencassie ,

thanks for your input! I really appreciate your efforts.

Just to make an example of a possible implementation.
E.g. when I have a flow like this:

```C#
var source = new DBSource<Type1>("sourceTable");
var rowTrans = new RowTransformation<Type1, Type2>( row => new Type2(row));
var dest = new DBDestination<type2>("destTable");

source.LinkTo(rowTrans).LinkTo(dest);
```

Basically I want to send data from a source into a destination ,while converting the object type in the transformation.

So how would error handling look like. If we want to store the exception together the faulted record , I would probably introduce some `ETLBoxError` class, that could look like this:

```C#
public sealed class ETLBoxError<T> {
    public T Record {get; set;}
    public Exception Error {get;set;}
}
```

Now let's assume we have an error in the source, and that we we want to store the record itself in a database which contains the error data, but additionally an extra column to store the exception message. We would need something like this

```
class Type1AndException : Type1 {
   public string ErrorMessage;
}

var errorsFromSource = new DBSource<Type1AndException>("ErrorSource");
var getType1AndException = new RowTransformation<ETLBoxError<Type1>, Type1AndException>
     ,  row => new Type1AndException() {
           SomeProperty = row.SomeProperty,
           ErrorMessage = row.Error.ToString() 
      });

source.LinkErrorsTo(getType1AndException).LinkTo(errorsFromSource);
```

If I want to store the errors from the DBDestination from the flow also in some database table, I would need a new class for `Type2` and a new error flow like the one above, but pointing to a table that could hold the data for Type2.

And for the `RowTransformation` (which would give me an error object like the DBDestination, so `ETLBoxError<Type1>`, I would probably need the exact same flow like above. But to reuse the flow, I would need to add a MergeJoin before, because a RowTransformation accepts only one input. 

This looks a little bit complicated - the error handling needs a lot more code than the actual data flow...

To wrap it up: adding an error output for every component should is feasible. But the question is: What kind of object should the error output "throw out" - my first thought was that it is a new object containing the faulted record alongside the exception, but to use it in a subsequent data flow and store the data somewhere could be really complicated. So I am actually looking for some better solution here - any ideas?

**Comment by saad749** [comment]
@roadrunnerlenny 

If I understand the issue correctly, how about storing the erroneous record in a serialized format as a string.

public sealed class ETLBoxError {
    public string Record {get; set;}
    public Exception Error {get;set;}
}

Maybe an additional property can be added for class type.

**Comment by roadrunnerlenny** [comment]
Hi @saad749 ,
thanks for participating in this discussion.
I like the idea of have the data record serialized into a string. So the `LinkErrorTo()` method could be linked to any components of Type `ETLBoxError`, and whoever is interested in the data could de-serialize the string or check the Exception.

But that would mean that the data object does need to provide a functionality to have it serialized. On first thought, this could be by simply overwriting the `ToString()` method, but perhaps it will be better to have some kind of Json serialization happening? Any thougths on this?

**Comment by saad749** [comment]
Yes, I think it should be json serialization.

**Comment by bencassie** [comment]
Hi,

Just wanted to say that I've been using your draft implementation and it is working quite well for me, no issues so far.

Sorry it has taken time for me to reply.

Ben

**Comment by roadrunnerlenny** [comment]
Hi Ben,

Thanks for your feedback! Good to know that you were able to test it, and that it works! Are you satisfied wit the approach itself?

I hope I will find the time to implement it for every task soon...

**Comment by roadrunnerlenny** [comment]
Hi all,

I just released version 1.7.7 (https://www.nuget.org/packages/ETLBox/1.7.7) which now enables you to redirect erroneous records within your flow. See the updated documentation here:
https://etlbox.net/articles/dataflow/dataflow_linking_execution.html#linking-errors

Thank you all for your support!
